src/kvstore/global_kvs.c

```diff
@@ -0,0 +1,131 @@
+/*
+ * Filename : global_kvs.c
+ * Description : 
+ * 
+ * Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ */
+
+#include <errno.h>
+#include <string.h>
+#include "kvstore.h"
+#include "global_kvs.h"
+
+/* Global definitions */
+kvs_idx_fid_t *g_fid = NULL;
+static struct kvstore *g_kvstore;
+struct kvs_idx g_index;
+extern bool kvs_init_done;
+
+/* Set an index mentioned by fid_str as global kvstore*/
+int nsal_set_global_kvs(const char *fid_str)
+{
+	int rc = 0;
+
+	if (g_fid != NULL) {
+		log_err("Global KVS is already set \n");
+		rc = -EEXIST;
+		goto out;
+	}
+	if (!kvs_init_done) {
+		log_err("kvs_init is not done \n");
+		rc = -EINVAL;
+		goto out;
+	}
+	g_fid = malloc(sizeof(kvs_idx_fid_t));
+	if (g_fid == NULL) {
+		log_err("Unable to allocate memory \n");
+		goto out;
+	}
+	rc = kvs_fid_from_str(fid_str, g_fid);
+	if (rc) {
+		log_err("Failed to get fid \n");
+		free(g_fid);
+		goto out;
+	}
+	g_kvstore = kvstore_get();
+	rc = kvs_index_open(g_kvstore, g_fid, &g_index);
+	if (rc) {
+		log_err("Failed to open the index \n");
+		free(g_fid);
+		goto out;
+	}
+
+out:
+	return rc;
+}
+
+/* Memory allocation for key */
+int nsal_global_alloc(void **ptr, size_t size)
+{
+	return kvs_alloc(g_kvstore, ptr, size);
+}
+
+/* free up the allocated memory */
+void nsal_global_free(void *ptr)
+{
+	kvs_free(g_kvstore, ptr);
+}
+
+/* Store KV pair in the global kvs */
+int nsal_global_set_kv(void *key, const size_t klen,
+                       void *value, const size_t vlen)
+{
+	int rc = 0;
+
+	if (g_fid == NULL) {
+		log_err("Global KVS is not set \n");
+		rc = -EINVAL;
+		goto out;
+	}
+	rc = kvs_set(g_kvstore, &g_index, key, klen, value, vlen);
+
+out:
+	return rc;
+}
+
+/* Get KV pair stored in the global kvs */
+int nsal_global_get_kv(void *key, const size_t klen,
+                       void **value, size_t *vlen)
+{
+	int rc = 0;
+
+	if (g_fid == NULL) {
+		log_err("Global KVS is not set \n");
+		rc = -EINVAL;
+		goto out;
+	}
+	rc = kvs_get(g_kvstore, &g_index, key, klen, value, vlen);
+	if(rc)
+		log_err("Unable to get the key \n");
```

Comment:
Question: In global_set_kv() and global_del_kv(), we are not checking the rc after kvs call. Is there any specific reason here?

Reply:
Actually I need to remove above , as I added it to root cause an issue I facing temporarily.
As I returning the rc, the caller to those global_xxx_kv should take required action.

Reply:
Okay, thanks.